xarray/core/resample.py

```diff
@@ -304,31 +307,36 @@ def _interpolate(self, kind='linear'):
         new_x = self._full_index.values.astype('float')
 
         arrs = []
-        # Apply the interpolation to each DataArray in our original Dataset
-        for v in self._obj.data_vars:
-            da = self._obj[v].copy()
-
-            x = da[self._dim].astype('float')
-            y = da.values
-            axis = da.get_axis_num(self._dim)
 
-            f = interp1d(x, y, kind=kind, axis=axis, bounds_error=True,
-                         assume_sorted=True)
+        # Prepare coordinates by dropping non-dimension coordinates along the
+        # resampling dimension.
+        # note: the lower-level Variable API could be used to speed this up
+        coords = OrderedDict()
+        if self._obj.data_vars:
+            var = list(self._obj.data_vars.keys())[0]
+            da = self._obj[var].copy()
```

Comment:
I'm not certain this version is correct anymore (so maybe restore what you had before).

If you want to go down the path of selecting coordinates once, you should iterate over `Dataset.coords` (not `DataArray.coords`) to get an `OrderedDict()`, and then merge it with the interpolated data variables all at once by calling the `Dataset()` constructor instead of `merge`.

Reply:
Ah, thanks for catching that. Pushing a few more commits that finish some of the documentation, and take this into consideration. I do want to do the coordinate selection all at once... but maybe we should go ahead and raise an issue to "enhance" this using Variable API, as you mention, if people run into issues.

Reply:
I think something like this would be work:
```python
old_times = self._obj[self._dim].astype(float)
new_times = self._full_index.values.astype(float)

data_vars = OrderedDict()
coords = OrderedDict()
for name, variable in self._obj.variables.items():
    if name in self._obj.coords:
        if name == self._dim:
            coords[self._dim] = self._full_index
        elif self._dim not in v.dims:
            coords[name] = variable
    else:
         if isinstance(variable.data, dask_array_type):
             raise TypeError('dask arrays not supported yet in resample().interpolate()')
         f = interp1d(old_times, variable.data, kind=kind, ....)
         interpolated = Variable(variable.dims, f(new_times), variable.attrs)
         data_vars[name] = interpolated
return Dataset(data_vars, coords)

Reply:
Oh that's much more succinct and nicer than my code. Hopefully it didn't take you much time to write! I made a few tweaks to it and dropped it into my code.

Reply:
did you push those commits yet? I'm not seeing it yet